source/binding/Bitstream.cpp

```diff
@@ -0,0 +1,664 @@
+                }
+
+                ASSERT(!dynamicSize || type.getArrayElementType()->isFixedSize());
+                dynamicSize = 0;
+            }
+        }
+        else {
+            const auto& fsua = ct.as<FixedSizeUnpackedArrayType>();
+            const auto& elem = fsua.elementType;
+            for (auto width = fsua.range.width(); width > 0; width--)
+                buffer.emplace(unpackBitstream(elem, iter, iterEnd, bit, dynamicSize));
+        }
+
+        return constContainer(ct, buffer);
+    }
+
+    if (type.isUnpackedStruct()) {
+        SmallVectorSized<ConstantValue, 16> buffer;
+        const auto& ct = type.getCanonicalType();
+        for (auto& field : ct.as<UnpackedStructType>().membersOfType<FieldSymbol>())
+            buffer.emplace(unpackBitstream(field.getType(), iter, iterEnd, bit, dynamicSize));
+
+        return constContainer(ct, buffer);
+    }
+
+    // TODO: classes
+    return nullptr;
+}
+
+ConstantValue Bitstream::evaluateCast(const Type& type, const ConstantValue& value,
+                                      SourceRange sourceRange, EvalContext& context,
+                                      bool isImplicit) {
+    auto srcSize = value.bitstreamWidth();
+    std::size_t dynamicSize = 0;
+    if (!isImplicit) { // Explicit bit-stream casting
+        dynamicSize = bitstreamCastRemainingSize(type, srcSize);
+        if (dynamicSize > srcSize) {
+            context.addDiag(diag::ConstEvalBitstreamCastSize, sourceRange)
+                << value.bitstreamWidth() << type;
+            return nullptr;
+        }
+    }
+    else { // implicit streaming concatenation conversion
+        auto targetWidth = type.bitstreamWidth();
+        if (targetWidth < srcSize) {
+            if (type.isFixedSize()) {
+                context.addDiag(diag::BadStreamSize, sourceRange) << targetWidth << srcSize;
+                return nullptr;
+            }
+            dynamicSize = srcSize - targetWidth;
+        }
+    }
+
+    SmallVectorSized<const ConstantValue*, 8> packed;
+    packBitstream(value, packed);
+
+    bitwidth_t bitOffset = 0;
+    auto iter = packed.cbegin();
+    auto cv = unpackBitstream(type, iter, packed.cend(), bitOffset, dynamicSize);
+    ASSERT(!dynamicSize);
+    if (isImplicit && iter == packed.cend())
+        ASSERT(bitOffset == 0);
+    else {
+        ASSERT(bitOffset == ((*iter)->isInteger() ? (*iter)->integer().getBitWidth()
+                                                  : (*iter)->str().length() * CHAR_BIT));
+        ASSERT(iter != packed.cend() && ++iter == packed.cend());
+    }
+    return cv;
+}
+
+bool Bitstream::canBeTarget(const StreamingConcatenationExpression& lhs, const Expression& rhs,
+                            const BindContext& context) {
+    if (rhs.kind != ExpressionKind::Streaming && !rhs.type->isBitstreamType()) {
+        context.addDiag(diag::BadStreamType, rhs.sourceRange) << std::string("source") << *rhs.type;
```

Comment:
I've been avoiding parameterizing diagnostic messages using English words like this (unless they are already language keywords), since it makes it much harder in the future for someone to translate the diagnostics into another language. I don't know if that would ever actually happen, but it's not much harder to define a few diagnostics instead.

Reply:
The diag::BadStreamType have 3 categories for "expression", "target" and "source". I will change it into 3 different diagCodes without the expression/target/source parameter, as you suggested.